### ISSUE TITLE
When and how to use on_commit_task in djhuey?

### ISSUE THREAD

**Issue by simkimsia** [coleifer/huey#805]
I'm sorry. I tried to find documentation on this `on_commit_task` but I cannot find it.

![image](https://github.com/coleifer/huey/assets/245021/c499c8d1-eb04-4dca-8723-8857d30271b8)

I did ask chatGPT for an explanation and this is what it gave me. 

Is the chatGPT correct or did it make any mistakes?

## ChatGPT answer

The `on_commit_task` decorator provided by djhuey allows you to defer the enqueuing of a task until after a database transaction has been successfully committed. This is particularly useful when you want to ensure that the task only runs if the database changes have been successfully saved, avoiding potential inconsistencies.

Here’s a detailed breakdown and an example to help you understand how and when to use it.

### Understanding `on_commit_task`

1. **Decorator Function**: The `on_commit_task` function is a decorator factory that returns a decorator.
2. **Task Wrapper**: It wraps the provided function `fn` in a Huey task (`task_wrapper`).
3. **Transaction Commit Hook**: It registers a callback (`enqueue_on_commit`) with Django’s transaction management system to enqueue the task after the current database transaction commits successfully.
4. **Result Handle**: It returns a Huey result handle immediately, even though the task may not be enqueued until later.

### Example Usage

Imagine you have a function that updates some data in the database and you want to send a notification only if the data is successfully updated and committed.

#### Step-by-Step Example

1. **Define the function**:
   Define a function `update_data` that performs some database operation.
2. **Wrap with `on_commit_task`**:
   Use the `on_commit_task` decorator to ensure the task is enqueued only after the database transaction commits.
3. **Usage in a View or Another Function**:
   Call the decorated function in your view or any other function where you handle database transactions.

#### Code Example

```python
from django.db import transaction
from huey.contrib.djhuey import task, on_commit_task
from myapp.models import MyModel
from myapp.notifications import send_notification

# Define the database operation function
def update_data(pk, new_value):
    instance = MyModel.objects.get(pk=pk)
    instance.value = new_value
    instance.save()

# Create a regular task to send a notification
@task()
def send_notification_task(pk):
    instance = MyModel.objects.get(pk=pk)
    send_notification(f"Data updated for {instance}")

# Create an on_commit_task to update data and then send a notification
@on_commit_task()
def update_data_and_notify(pk, new_value):
    update_data(pk, new_value)
    send_notification_task(pk)

# Usage in a Django view
from django.http import JsonResponse

def my_view(request):
    pk = request.POST['pk']
    new_value = request.POST['new_value']
    
    with transaction.atomic():
        # This will ensure the task is enqueued only if the transaction commits
        update_data_and_notify(pk, new_value)
    
    return JsonResponse({'status': 'success'})
```

### Explanation

1. **Define `update_data`**:
   This function updates a database record with a new value.
2. **Define `send_notification_task`**:
   This regular Huey task sends a notification based on the updated data.
3. **Define `update_data_and_notify`**:
   This uses `on_commit_task` to ensure `update_data` and `send_notification_task` are executed only if the database transaction commits successfully.
4. **Use in a View**:
   In the view, use `transaction.atomic()` to wrap the database operations in a transaction. The `update_data_and_notify` function is called within this transaction, ensuring it will only enqueue the task if the transaction is successful.

### Key Points

- **Atomic Transactions**: Using `transaction.atomic()` ensures that all database operations within the block are treated as a single transaction.
- **Deferred Task Execution**: The task is only enqueued after the transaction is committed, preventing tasks from running with uncommitted or potentially rolled-back data.
- **Immediate Result Handle**: The function returns a result handle immediately, allowing the calling code to proceed without waiting for the transaction to complete.

This approach helps maintain data consistency and ensures that tasks are only executed when the related database operations are successfully committed.

**Comment by coleifer** [comment]
You would use `on_commit_task()` just like you would use the regular `task()` - although take care because it doesn't behave the same as regular `task()` (it does not provide the same facilities, such as scheduling, etc).

```python

@on_commit_task()
def do_work(obj_id):
    obj = Obj.objects.get(pk=obj_id)
    # Do some work.

def some_view(...):
    obj = Obj.objects.create(...)
    do_work(obj.id)  # This job will be enqueued *after* the transaction commits.
    return response
```

The main purpose of this decorator is when you need to specifically enqueue a task *after* the active transaction commits, for example in a view. If you don't use `on_commit_task()` there is a possibility of the task being enqueued and executed by the consumer before the transaction has committed (and thus the task will fail because the database row does not exist, e.g.).

**Comment by simkimsia** [comment]
@coleifer your example doesn't use transaction despite you mentioning how the on-commit-task is only enqueued AFTER the active transaction is complete.

So if I were to adapt ChatGPT example of sending an email notification after the update is completed

The send_email is the one with on_commit_task decorator yes?

**Comment by coleifer** [comment]
I've updated the code sample to use a transaction on the view (`@transaction.atomic`).
